pulse2percept/models/beyeler2019.py

```diff
@@ -904,6 +916,42 @@ def plot(self, use_dva=False, style='hull', annotate=True, autoscale=True,
             ann.set_yticks([])
         return ax
 
+class TorchAxonMapSpatial(TorchBaseModel):
+    def __init__(self, p2pmodel): # implant parameter currently not used
+        super().__init__(p2pmodel)
+
+        # Check for model validity
+        if isinstance(p2pmodel, AxonMapModel) or not isinstance(p2pmodel, AxonMapSpatial):
+            raise ValueError("Must pass in a valid AxonMapModelSpatial")
+        if p2pmodel.engine != 'torch':
+            raise ValueError("Engine Selection Conflict : Constructing TorchAxonMapSpatial with engine", p2pmodel.engine)
+
+        self.rho = torch.tensor(p2pmodel.rho, device=p2pmodel.device, dtype=torch.get_default_dtype())
+        self.axon_contrib = torch.tensor(p2pmodel.axon_contrib, device=p2pmodel.device, dtype=torch.get_default_dtype()) # to(self.device)
+        self.axlambda = torch.tensor(p2pmodel.axlambda, device=p2pmodel.device, dtype=torch.get_default_dtype())
+        self.thresh_percept = p2pmodel.thresh_percept
+        self.percept_shape = p2pmodel.grid.shape
+
+        
+    
+    def forward(self, stim, e_locs):
+        
+        # I_axon(x,y;p, lambda) = I_score(x,y; p) * exp(-( (x-x_soma)**2 + (y-y_soma)**2) / (2 * lambda**2))
+        #                       = amp * gauss
+        # gauss = self.axon_contrib[:,:,2,None] * torch.exp(-d2_el / (2*self.rho**2))
+        
+        amp = torch.tensor(stim[:, ::1], dtype=torch.get_default_dtype()) # I_score
+        
+        d2_el = ((self.axon_contrib[:, :, 0, None] - e_locs[:,0])**2 + (self.axon_contrib[:, :, 1, None] - e_locs[:,1])**2)[None, :, :, :] # (x-x_soma)**2 + (y-y_soma)**2
+        gauss = self.axon_contrib[:,:,2,None] * torch.exp(-d2_el / (2*self.rho**2))
+        intensities = amp[:, None, None, :] * gauss
+        intensities = torch.sum(intensities, axis=-1)
+        intensities = torch.max(intensities, dim=-1).values
```

Comment:
I think we need one more line to match previous behavior, basically clip anything beneath self.thresh_percept to be 0

Reply:
intensities = torch.where(intensities > self.thresh_percept, intensities, 0)